src/engine/Pro_hooks.ml

```diff
@@ -0,0 +1,84 @@
+(* Iago Abal
+ *
+ * Copyright (C) 2023-2024 Semgrep Inc.
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+(*****************************************************************************)
+(* Prelude *)
+(*****************************************************************************)
+(* Gather all pro hooks in one place.
+ *
+ * alt: maybe we could move this file in src/core/, and actually define
+ * the hooks there, but some hooks depends on types outside src/core/
+ * so simpler to keep the hook spreaded, but at least reset them here.
+ *
+ * alt: this could could be moved to semgrep-pro. It's only used by
+ * Globals.reset() which itself is not used.
+ *)
+
+(*****************************************************************************)
+(* Types and constants *)
+(*****************************************************************************)
+
+type pro_hook_ref =
+  | Pro_hook_ref : 'a option ref -> pro_hook_ref
+      (** 'pro_hook_ref' is a GADT, this allows us to do an "existential"
+       * quantification over `'a`. That is, you can take any hook with
+       * type '... option ref' and wrap it as a 'pro_hook_ref'. Of course,
+       * when you pattern match on one of this, you cannot make any assumptions
+       * on what that `'a` is, but it is enough to reset the hooks.
+       * So we can enumerate Pro hooks in 'pro_hooks_refs' and then write
+       * 'reset_pro_hooks' and 'save_pro_hooks_and_reset' based on that list.
+       *)
+
+(* TODO: a steps-mode rule to ensure we have all the pro hooks.
```

Comment:
@emjin steps mode!!! another use case.

Reply:
Do typed metavariables work for Ocaml? I'm just trying to figure out how to write the part where we detect hooks. (Or we could just assume that they're prefixed with `hook_` lol)

Reply:
If you write a rule that detects pro hooks I'll turn it multistep for you